File: DiffieHellman.py
"""
By BingLi224

01:11 THA 01/02/2019

Diffie Hellman
"""

last_prime_check_value = 2
prime_set = [ 2 ]

def update_prime_set ( value ) :
    global last_prime_check_value, prime_set
    if value > last_prime_check_value :
        ## check all numbers in the current target
        for v in range ( last_prime_check_value + 1, value ) :
            b_prime = True ## presume to be prime
            for p in prime_set :
                if v % p == 0 :
                    ## not prime
                    b_prime = False
                    break
                
            if b_prime :
                prime_set += [ v ]
            
            last_prime_check_value = value

def is_prime ( value ) :
    global prime_set
    update_prime_set ( value )
    return value in prime_set    

def is_co_prime ( v1, v2 ) :
    global prime_set
    update_prime_set ( v1 if v1 > v2 else v2 )
    
    for p in prime_set :
        if p > v1 or p > v2 :
            return True
        if v1 % p == 0 and v2 % p == 0 :
            return False
            
    return True

class DiffieHellman :
    base = 0
    mod = 0
    
    def __init__ ( this, base, mod ) :
        this.base = base
        this.mod = mod
    
    def encrypt ( this, value, base = None ) :
        if base is None :
            base = this.base
            
        code = 1
        for i in range ( 1, value ) :
            code *= base
            if code >= mod :
                code %= mod

        return code

###############################################################################
## test

base = 443
mod = 743

print ( "public prime\t: " + str ( mod ) )
print ( "public base\t: " + str ( base ) )

print ( "Alice's secret\t: " )
try :
    v1 = int ( input ( ) ) # 241
except :
    v1 = 241
    print ( "\tDefault: " + str ( v1 ) )
    
print ( "Bob's secret\t: " )
try :
    v2 = int ( input ( ) ) # 87
except :
    v2 = 87
    print ( "\tDefault: " + str ( v2 ) )
    
## test if co-prime
print ( "co-prime(" + str ( v1 ) + "," + str ( v2 ) + ") = " + str ( is_co_prime ( v1, v2 ) ) )

dh = DiffieHellman ( base, mod )
c1 = dh.encrypt ( v1 )
c2 = dh.encrypt ( v2 )

print ( "Alice's cipher\t: " + str ( c1 ) )
print ( "Bob's cipher\t: " + str ( c2 ) )
print ( "Alice's decrypt\t: " + str ( dh.encrypt ( v1, base = c2 ) ) )
print ( "Bob's decrypt\t: " + str ( dh.encrypt ( v2, base = c1 ) ) )

File: DiffieHellman.kt

/**
 * @author BingLi224
 * @version 2019.01.29
 *
 * Diffie-Hellman
 */
 
fun main(args: Array<String>) {

    // base ^ v (** mod)
    val base = 443L
    val mod = 743L

    println ( "public prime\t: $mod")
    println ( "public base\t: $base")

    val v1 = 241L //args [0].toLong ( )
    
    println ( "Alice's secret\t: $v1")

    val v2 = 87L //args [1].toLong ( )
    
    println ( "Bob's secret\t: $v2")

    // test if co-prime    
    println ( "co-prime($v1, $v2) = ${DiffieHellman.isCoPrime (v1, v2)}")
    
    val dh = DiffieHellman ( base, mod )
    val c1 = dh.encrypt ( v1 )
    val c2 = dh.encrypt ( v2 )
    
    println ( "Alice's cipher\t: $c1")
    println ( "Bob's cipher\t: $c2")
    
    println ( "Alice's decrypt\t: ${dh.encrypt(v1, base=c2)}")
    println ( "Bob's decrypt\t: ${dh.encrypt(v2, base=c1)}")
}

class DiffieHellman ( val base : Long, val mod : Long ) {
    companion object {
        var lastPrimeCheckValue = 2L
        val primeSet = mutableSetOf <Long> (2L)

        fun updatePrimeSet ( value : Long ) {
            if ( value > lastPrimeCheckValue ) {
                
                
                // check all numbers to currect target
                for ( v in lastPrimeCheckValue + 1 .. value ) {
                
                    var bPrime = true // presume to be prime
                    testRem@ for ( p in primeSet ) {
                        if ( v.rem ( p ) == 0L ) {
                            // not prime
                            bPrime = false 
                            break@testRem
                        }
                    }
                    
                    if ( bPrime )
                        primeSet.add ( v )
                }
                
                lastPrimeCheckValue = value
            }
        }
        
        fun isPrime ( value : Long ) : Boolean {
            updatePrimeSet ( value )
            return primeSet.contains (value)
        }
        
        fun isCoPrime ( v1 : Long, v2 : Long ) : Boolean {
            updatePrimeSet ( if ( v1 > v2 ) v1 else v2 )
            for ( p in primeSet ) {
                if ( p > v1 || p > v2 )
                    return true 
                if ( v1.rem ( p ) == 0L && v2.rem ( p ) == 0L )
                    return false 
            }
            
            return true 
        }
    }
    
    fun encrypt ( value : Long, base : Long = this.base ) : Long {
        var code = 1L
        for ( i in 1L .. value ) {
            code *= base
            if ( code >= mod )
                code = code.rem ( mod )
            //println ("v=$value b=$base c=$code")
        }
        
        return code
    }
}
